경북대 FE 장세현 3단계 - 테마 상품 목록 페이지 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
3단계 과제 - '테마 상품 목록 페이지 구현하기' 내용을 PR 요청 드립니다!

---

### 과제 진행 사항

#### 목표

- `Intersection Observer API`를 사용하여 페이지네이션을 구현
- 참고 URL을 참고하여 테마 상품 목록 페이지를 구현
- axios외에 별도의 fetch 관련 라이브러리를 사용하지 않고 구현
- Suspense, ErrorBoundary 사용하지 않기

#### 요구사항

- 테마 상품 목록 페이지
  - 선물하기 홈 > 선물 테마 섹션의 아이템을 클릭하면 테마 상품 목록 페이지로 연결
  - `/api/themes/:themeId/info` API를 사용하여 선물 테마 섹션의 히어로 영역을 구현
  - 만약 테마 정보를 가져오는 API가 404 에러가 발생하면 선물하기 홈 페이지로 연결
  - `/api/themes/:themeId/products` API를 사용하여 상품 리스트를 구현
  - 무한 스크롤 기능을 구현
  - 상품 리스트가 없으면 빈 페이지를 보여주기

- 본인만의 기준으로 일관된 코드 작성
- 기능 단위로 나누어 커밋
- 재사용을 고려하여 컴포넌트, 비즈니스 로직을 구현

#### 추가 진행 사항

- emotion 의 ThemeProvider를 사용하고 있음에도 각 코드에서 import 해서 theme을 사용하던 부분을 props로 받아 사용하거나 useTheme()을 사용하도록 본 취지에 맞게 전체 리팩토링 진행하였습니다.

---

### 질문 사항

- 예시 페이지를 보니, 제가 만든 것과 달리 데이터 로딩 속도가 크게 느린 것은 아닐 텐데도, 약간의 지연 후 데이터를 받아오는 모습을 확인할 수 있었습니다.
반면, 제가 구현한 현재 상태에서는 로딩이 너무 빠르게 끝나 거의 깜빡이는 수준으로 표시되기 때문에, 일부러 코드를 수정하지 않으면 로딩 화면을 제대로 보기 어려운 상황입니다.
그래서 “이게 사용자 입장에서 보다 부드러운 화면 전환을 위해 의도적으로 로딩을 길게 설정한 걸까?“라는 의문이 들었습니다.
실제로 여러 사이트에서 로딩 화면이나 스켈레톤 UI를 자주 사용하는 것을 보면, 단순히 속도 문제 때문이 아니라 의도적으로 시간을 늘려 사용자 경험을 개선하려 한건지, 이 부분에 대해서 여쭤보고 싶습니다.

SeHIgh committed:
docs(README): STEP 3 과제 목표 및 요구사항 추가

SeHIgh committed:
refactor(theme): 카테고리 섹션을 요구사항에 따라 테마 섹션으로 명칭 변경 및 추가 리팩토링 진행

- 컴포넌트 명칭 변경 ('Category' -> 'ThemesSection')
- theme 도메인의 기능 세분화로 features 로 새롭게 통합관리를 위해 이전
  (components/themes -> features/themes/components)

SeHIgh committed:
feat(themes): 테마 상품 목록 페이지 API 구현

- Router에 테마 상품 목록 페이지 경로 추가
- API 엔드포인트에 테마 정보 및 상품 목록 API 추가
- 테마 상세 정보 및 상품 목록을 불러오는 서비스 함수 구현
- 테마 관련 타입 정의 추가

SeHIgh committed:
feat(themes): HeroSection 구현

- 테마 정보를 기반 내용을 표시하는 HeroSection 컴포넌트 구현

SeHIgh committed:
feat(hooks): 특정 에러 상태에 대한 관리를 위해 useFetch 훅에 에러 상태 코드 처리 추가

- AxiosError 타입을 사용하여 에러 발생 시 상태 코드를 fetchState에 포함하도록 추가

SeHIgh committed:
feat(themes): 테마 상품 목록 페이지 및 관련 컴포넌트 추가

- 테마 상품 목록 페이지 구현
- useFetch 훅을 사용하여 테마 정보 fetch 기능 추가
- 로딩 및 에러 처리 로직 구현

SeHIgh committed:
feat(hooks): 페이지네이션을 위한 커스텀 훅 usePagination 추가

- 페이지네이션 기능을 구현하기 위한 usePagination 훅 추가
- 데이터 페칭, 로딩 상태, 에러 처리 및 초기화 기능 포함

SeHIgh committed:
feat(themes): 테마 상품 목록 페이지 기능 개선 및 컴포넌트 추가

- 테마 페이지에 상품 리스트 컴포넌트 추가
- 기존 ProductItem 컴포넌트를 개선하여 재사용
  - 인덱스가 없을 경우 처리 로직 개선 (순위가 필요하지 않은 경우를 위해)
  - 상품 가격의 가시성을 위해 toLocaleString() 추가
- HeroSection 및 ProductList 컴포넌트 임포트 추가

SeHIgh committed:
feat(themes): 상품 목록 컴포넌트에 무한 스크롤 기능 추가

- ProductList 컴포넌트에 무한 스크롤 기능 구현
- 무한 스크롤을 위한 Intersection Observer 커스텀 훅 구현
- 조건부 렌더링 로직 개선

SeHIgh commented on src/api/types/theme.ts:
```diff
@@ -1,6 +1,24 @@
+
+// * 테마별 상품 목록 응답 타입
+export interface ThemeProductListResponse {
+  list: Product[]
```
이전에 랭킹 상품 목록 섹션에서 구현한 Product 타입을 재사용 하였습니다.

SeHIgh commented on src/features/product/components/trend/ProductItem.tsx:
```diff
@@ -17,7 +17,7 @@ export const ProductItem = ({ product, index }: { product: Product; index: numbe
   return (
     <ProductItemContainer onClick={handleProductClick}>
-      <ProductRank rank={productRank}>{productRank}</ProductRank>
+      {productRank && <ProductRank rank={productRank}>{productRank}</ProductRank>}
```
기존에 구현한 상품 아이템 컴포넌트를 재사용하기 위해
순위 표시가 불필요한 경우를 위해 렌더링 조건 등을 추가하였습니다.

SeHIgh replied:
기존에 구현한 상품 아이템 컴포넌트를 재사용하기 위해
순위 표시가 불필요한 경우를 위해 렌더링 조건 등을 추가하였습니다.

SeHIgh commented on src/features/themes/components/ProductList.tsx:
```diff
@@ -0,0 +1,112 @@
+          {products.map((product) => (
+            <ProductItem key={product.id} product={product} />
+          ))}
+          {canObserve && <div ref={loaderRef} style={{ height: 0 }} />}
```
하단에 안보이는 div를 추가하여 해당 div를 기점으로 무한스크롤이 되도록 구현하였습니다

SeHIgh commented on src/hooks/useFetch.ts:
```diff
@@ -28,10 +30,17 @@ export const useFetch = <T>(fetcher: () => Promise<T>, deps: unknown[] = []) =>
+        }
         if (isMounted) {
-          setFetchState({ isLoading: false, isError: true, data: null })
+          setFetchState({ isLoading: false, isError: true, data: null, errorStatus })
```
특정 에러에 대한 처리를 할 수 있도록 useFetch 커스텀 훅을 개선하였습니다.

SeHIgh commented on src/hooks/usePagination.ts:
기존 useFetch() 훅으로는 한계가 있어서 페이지네이션 전용 커스텀 훅을 추가로 구현하였습니다.

SeHIgh committed:
refactor(style): styled-components에서 theme 사용 방식 통일

- 모든 styled-components에서 theme을 props로 받아 사용하는 방식으로 변경
- 코드 가독성 및 유지보수성 향상

Leejin-Yang commented on src/features/product/components/trend/ProductItem.tsx:
```diff
@@ -29,12 +31,15 @@ export const ProductItem = ({ product, index }: { product: Product; index: numbe
             <ProductDiscountRate>{product.price.discountRate}%</ProductDiscountRate>
             <ProductBasicPrice style={{ textDecoration: 'line-through' }}>
-              {product.price.basicPrice}
+              {product.price.basicPrice.toLocaleString()}
```
👍

Leejin-Yang commented on src/hooks/usePagination.ts:
```diff
@@ -0,0 +1,66 @@
+  const [isLoading, setIsLoading] = useState(false)
+  const [isError, setIsError] = useState(false)
+
+  // * fetchMore가 항상 최신 cursor를 사용하기위해
+  const cursorRef = useRef(cursor)
+  cursorRef.current = cursor
```
ref만 사용했을 때와 어떤 차이가 있을까요?

SeHIgh replied:
저도 이부분은 gpt도움을 조금 받았는데, 
제가 이해하기로는 상태를 변경하는 부분이 useCallback 내부에 위치하기 때문에 최신 cursor값이 제대로 반영이 되지 않을 수 있는 부분이 있어 항상 최신 cursor를 따로 저장하고 이를 사용하도록 해서 문제를 방지하는 용도로 추가한 부분이라 생각합니다.

Leejin-Yang replied:
useState와 useRef의 차이에 대해서도 학습해보면 좋아보여요 😊

Leejin-Yang commented on src/hooks/useFetch.ts:
```diff
@@ -28,10 +30,17 @@ export const useFetch = <T>(fetcher: () => Promise<T>, deps: unknown[] = []) =>
+        if (typeof error === 'object' && error !== null && 'response' in error) {
+          const err = error as AxiosError
+          errorStatus = err.response?.status
+        }
```
instanceof를 활용해 조건문을 간결하게 할 수 있어보여요 😊

```suggestion
        if (error instanceof AxiosError && error.response) {
          const errorStatus = error.response.status
          setFetchState({ isLoading: false, isError: true, data: null, errorStatus })
        }
```

Leejin-Yang commented on src/hooks/usePagination.ts:
```diff
@@ -0,0 +1,66 @@
+  const [cursor, setCursor] = useState(initialCursor)
+  const [hasMore, setHasMore] = useState(true)
+  const [isLoading, setIsLoading] = useState(false)
+  const [isError, setIsError] = useState(false)
```
이 상태들도 useFetch와 같이 하나의 상태로 관리할 수 있겠네요 😊

Leejin-Yang commented on src/pages/Themes.tsx:
```diff
@@ -0,0 +1,85 @@
+  } = useFetch<ThemeInfo>(() => fetchThemeInfo(themeId), [themeId])
+
+  // * 상품 리스트 페이지네이션 커스텀 훅
+  const {
+    list: productList,
+    hasMore,
+    isLoading: isProductsLoading,
+    isError: isProductsError,
+    fetchMore,
+    resetAndFetch,
+  } = usePagination<Product>({
+    fetcher: useCallback((cursor, limit) => fetchThemeProducts(themeId, cursor, limit), [themeId]),
+    initialCursor: 0,
+    limit: 20,
+    deps: [themeId],
+  })
```
지금처럼 페이지에서 모든 데이터를 페칭해서 props로 전달하는 구조와 이 로직은 ProductList에서 처리하는 구조가 있어 보이는데요, 둘은 어떤 차이점이 있을까요? 리렌더링 측면에서 생각해볼 수 있겠네요 😊

SeHIgh replied:
저는 일단 이렇게 구성한 이유가 
이번 과제 예시 페이지로 보면, 짧은 순간이지만 히어로 섹션이 먼저 로딩이 되고, 그 다음에 상품 목록이 로딩이 되는 것을 보고
두 데이터를 페칭을 한 곳에서 수행해야 데이터 흐름을 적절히 제어하여 의도에 맞게 구현될 수 있다 생각하였기 때문입니다.

그리고 추가로 멘토님께서 말씀해주신 리렌더링 측면에서 생각해보면, 
상위 상태 변화가 있을때만 하위가 리렌더링 되기도 하고,
불필요한 중복 페치도 줄일 수 있다 생각합니다!

Leejin-Yang replied:
많이 고민해주시고 작성해주셨네요 👍
아래 작성한 코멘트는 이번 미션 외적인 부분이니, 아직 참고만 해주세요 😊

지금의 구조에서 흐름을 봤을 때

- Themes 렌더링 (로딩중이므로 로딩 화면)
- ThemeInfo 페칭, **ProductList 페칭**
- ThemeInfo 성공적으로 가져오면 페이지 렌더링
  - **ProductList 통신 상태에 따라 컴포넌트 렌더링**

이런 흐름으로 갈 수 있다고 보여요
만약 productList를 ProductList 안에서 요청을 한다면

- Themes 렌더링 (로딩중이므로 로딩 화면)
- ThemeInfo 페칭
- ThemeInfo 성공적으로 가져오면 페이지 렌더링
  - 이때 ProductList 렌더링 (로딩중)
- **ProductList 페칭**

이런 흐름으로 갈 수 있겠네요
컴포넌트 렌더링 순서에 따라 각 로직이 실행되는 순서가 달라질 수 있어요
다만, ProductList가 Theme에서만 사용하는 컴포넌트가 아니라면 props로 넘겨줘야하는 구조를 가져가야 할수도 있지만요 😅

그리고 리렌더링 측면에서는 현재는 상품 리스트를 계속 페칭해오면 Themes 컴포넌트 전체가 리렌더링 되고 있어요. 그러나 이 로직을 ProductList 안으로 한다면 데이터 페칭이 일어날 때 ProductList 컴포넌트만 리렌더링 하도록 할 수 있어요

Themes 컴포넌트에서 productList 관련 상태값을 사용하지 않는다면, 불필요한 리렌더링 방지를 위해 로직을 옮기는 것을 생각할 수 있습니다 😊 관련해서는 개발자도구에서 profiler를 통해 확인해볼 수 있어요

Leejin-Yang commented on src/api/services/theme.ts:
```diff
@@ -1,9 +1,27 @@
+  limit = 10,
+): Promise<ThemeProductListResponse> => {
+  const res = await apiClient.get<{ data: ThemeProductListResponse }>(
+    `${API_ENDPOINTS.THEMES.PRODUCTS(themeId)}?cursor=${cursor}&limit=${limit}`,
```
작성해주신 fetchProductRankList 처럼params 속성을 사용해도 좋아보여요 😊

Leejin-Yang reviewed (changes requested):
세현님 안녕하세요 👋 마지막 단계도 잘 진행해주셨네요 👍
기존 컴포넌트도 재사용해주시고, 기존 구조에서 잘 확장해주셨다고 생각이 들어요
마지막 단계 이렇게 보내드리기 아쉬워서... 몇가지 이야기할 포인트에 코멘트 달아뒀어요
확인해주시고 다시 요청 보내주세요 😊

그리고 질문에 답변을 하자면,

> 그래서 “이게 사용자 입장에서 보다 부드러운 화면 전환을 위해 의도적으로 로딩을 길게 설정한 걸까?“라는 의문이 들었습니다.

이러한 개념에서도 자주 쓰일 수 있어요 👍 사용자 경험 측면에서 잘 접근해주셨네요
또한 콘텐츠에 따라 로딩시간을 의도적으로 길거나 짧게 잡을 수도 있어요
다만 속도가 늦어지면 사용자 경험적으로 좋지 않으니 이러한 부분도 고민해봐야하구요
현재는 가벼운 앱이고 로컬에서 실행하니 속도에 크게 신경쓰지 않아도 괜찮지만, 앞으로 여러 프로젝트를 진행하면서 고민해보셔도 좋아보여요. 관련해서 [아티클](https://web.dev/learn/performance/why-speed-matters?hl=ko) 하나 남겨둘게요 😊

SeHIgh committed:
refactor(hooks): useFetch 훅에서 AxiosError 처리 로직 개선

- instanceof 를 활용하여 조건문 개선
(AxiosError 인스턴스 확인 후 에러 상태 코드 추출 방식 수정)

SeHIgh committed:
refactor(hooks): usePagination 훅의 상태 관리 개선

- 페이지네이션 상태를 관리하기 위해 PaginationState 인터페이스 추가
- 상태를 객체로 통합하여 코드 가독성 향상
- 변경한 상태를 응용하여 fetchMore 및 resetAndFetch 함수에서 상태 업데이트 방식 개선

SeHIgh committed:
refactor(api): fetchThemeProducts 함수의 API 호출 방식 개선

- API 호출 시 쿼리 파라미터를 params 객체로 전달하도록 변경
- 코드 가독성 향상 및 유지보수성 개선

SeHIgh commented:
멘토님 상세한 피드백과 좋은 말씀 감사합니다!

멘토님의 피드백을 반영하여 코드를 수정 및 반영하였습니다!

1. instanceof를 사용하여 useFetch 훅에서 AxiosError 처리 로직을 개선하였습니다. 

2. usePagination 훅의 상태 관리를 하나의 객체로 통합 관리하도록 개선하고 그를 응용하여 상태변경도 개선해보았습니다.

3. fetchThemeProducts 함수의 API 호출 시, 쿼리스트링 대신 params 속성으로 cursor와 limit을 전달하도록 개선하였습니다.

Leejin-Yang commented on src/api/services/theme.ts:
```diff
@@ -21,7 +21,13 @@ export const fetchThemeProducts = async (
-    `${API_ENDPOINTS.THEMES.PRODUCTS(themeId)}?cursor=${cursor}&limit=${limit}`,
+    API_ENDPOINTS.THEMES.PRODUCTS(themeId),
+    {
+      params: {
+        cursor,
+        limit,
+      },
```
👍

Leejin-Yang reviewed (approved):
빠르게 피드백 반영해주셨네요 👍
이번 미션을 매우 잘 진행해주셔서, 저도 세현님을 통해 많이 배우고갑니다 😊
더 이야기하면 좋겠지만,, 아쉽게도 이번 미션은 여기서 마무리 해야겠네요
앞으로도 개발적인거나 외적인거나 궁금한점이 있다면 언제든 DM 남겨주세요
고생 많았어요 🙌
